### PR TITLE
[3.x] Improve Scene Tree Dock's Node filter (Allow multiple terms & more)

### PR DESCRIPTION
**Pull request by Mickeon**
Backport of #65352, #65932 and #65939, as well as some other filtering behavioral changes.

For the onlookers perhaps seeing this on a patch-note, here's the summary of what this PR features:

| Multiple filter terms & more | Special filters
| --- | --- |
| Gif2 | Gif1
| Spaces divide each search term. <br>Only Nodes matching **all** terms are displayed. <br>Parent Nodes with children matching the search are **darkened**. | A few special filters are available to narrow down the search: <br> - **Filter by Type** (accounts for inheritance); <br>- **Filter by Group**.

If you do not remember the special filters, or want to discover this feature, they are quickly available by:
| Right click | Middle click | Top-right menu |
| --- | --- | --- |
| image| middle Click | No, wait, this one is not in Godot 3... <br>What do I do?

| Warning | Tooltip |
| --- | --- |
| Warning | Tooltip |
| A warning is available as a tooltip when the special filter is not recognised. | A short description can be seen by hovering over the "Filter by" options |

Lastly, you may have noticed dark yellow-coloured names:
| | |
|--- | --- |
| image | image

~~Out of necessity, in this PR, this is the color of inherited children. 
Do note that this is actually **unlike** Godot 4.0, which uses a purely unfiltered warning yellow colour. However, I personally felt like that was too alienating to returning users.~~

### CODE REVIEW

**Comment by Mickeon** [comment]
This has been ready for review for quite a while, it would be nice for someone to have a look.

**Comment by akien-mga** [comment]
Would be worth rebasing as the last build was 1.5 years ago :)

**Comment by Mickeon** [comment]
Rebased to latest with surprisingly little problems at a glance. Let's see.

File: editor/scene_tree_dock.cpp [hunk 1]
```diff
@@ -3405,14 +3471,23 @@ SceneTreeDock::SceneTreeDock(EditorNode *p_editor, Node *p_scene_root, EditorSel
...
+
+	// The "Filter Nodes" text input above the Scene Tree Editor.
 	filter = memnew(LineEdit);
 	filter->set_h_size_flags(SIZE_EXPAND_FILL);
 	filter->set_placeholder(TTR("Filter nodes"));
 	filter_hbc->add_child(filter);
 	filter->add_constant_override("minimum_spaces", 0);
 	filter->connect("text_changed", this, "_filter_changed");
+	filter->connect("gui_input", this, "_filter_gui_input");
+	//filter->get_menu()->connect("about_to_show", this, "_update_filter_menu");
```
Comment (KoBeWi) [inline editor/scene_tree_dock.cpp:3484]: Why did you add a commented-out line?

**Review by KoBeWi — APPROVED** [review]
I tested the functionality and tried to find some obvious regressions; it works correctly.
I didn't do a super-deep code review, but it looks fine at a glance.

The new foreign node color looks rather washed-out though and blends with the background:
image
Might be worth it to make it more prominent, like in Godot 4:
image

File: editor/scene_tree_dock.cpp (hunk 1 above)
Comment (Mickeon) [reply editor/scene_tree_dock.cpp:3484]: Unsure, it's been a whole year 💦

**Comment by Mickeon** [comment]
I updated the PR to use the same colors as master. Godot 4 has been out for long enough at this point.

**Comment by akien-mga** [comment]
Thanks!
